prov/efa/src/rxr/rxr_pkt_type_req.h

```diff
@@ -0,0 +1,269 @@
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _RXR_PKT_TYPE_REQ_H
+#define _RXR_PKT_TYPE_REQ_H
+
+/*
+ * This file contain REQ packet type related struct and functions
+ * REQ packets can be classifed into 4 categories:
+ *    RTM (Request To Message) is used by message
+ *    RTW (Request To Write) is used by RMA write
+ *    RTR (Request To Read) is used by RMA read
+ *    RTA (Request To Atomic) is used by Atomic
+ *
+ * For each REQ packet type need to have the following:
+ *
+ *     1. a header struct
+ *     2. an init() function called by rxr_pkt_init_ctrl()
+ *     3. a handle_sent() function called by rxr_pkt_post_ctrl()
+ *     4. a handle_send_completion() function called by
+ *               rxr_pkt_handle_send_completion()
+ *     5. a proc() function called by
+ *               rxr_pkt_proc_req()
+ *
+ * Some req packet types are so similar that they can share
+ * some functions.
+ */
+
+/*
+ * Utilities shared by all REQ packets
+ *
+ *     Flags
+ */
+#define RXR_REQ_OPT_RAW_ADDR_HDR	BIT_ULL(0)
+#define RXR_REQ_OPT_CQ_DATA_HDR		BIT_ULL(1)
+#define RXR_REQ_MSG			BIT_ULL(2)
+#define RXR_REQ_TAGGED			BIT_ULL(3)
+#define RXR_REQ_RMA			BIT_ULL(4)
+
+/*
+ *     Utility struct and functions for
+ *             REQ packet types
+ */
+struct rxr_req_opt_raw_addr_hdr {
+	size_t addr_len;
+	char raw_addr[32];
+};
+
+struct rxr_req_opt_cq_data_hdr {
+	int64_t cq_data;
+};
+
+extern size_t RXR_REQ_HDR_SIZE_LIST[];
+
+void rxr_pkt_read_req_hdr(struct rxr_pkt_entry *pkt_entry,
+			  struct rxr_rx_entry *rx_entry);
+
+size_t rxr_pkt_req_total_len(struct rxr_pkt_entry *pkt_entry);
+
+size_t rxr_pkt_req_max_data_size(struct rxr_ep *ep, fi_addr_t addr, int pkt_type);
+
+/*
+ * Structs and funcitons for RTM (Message) packet types
+ * There are 4 message protocols
+ *         Eager message protocol,
+ *         Medium message protocol,
+ *         Long message protocol,
+ *         Read message protocol (message by read)
+ * Each protocol employes two packet types: non-tagged and tagged.
+ * Thus altogether there are 8 RTM packet types.
+ */
+
+/*
+ *   Utility structs and functions shared by all
+ *   RTM packet types
+ */
+struct rxr_rtm_base_hdr {
+	uint8_t type;
+	uint8_t version;
+	uint16_t flags;
+	uint32_t msg_id;
```

Comment:
I don't follow - send/recv doesn't always require reordering. We turn reordering off if the app doesn't need it for all message types.

Reply:
Ah, I forgot that, but that is uncommon and IMO does not warrant a branching code. Also, removing msg_id would only affect size of rxr_eager_msgrtm_hdr. The other 3 header rxr_eager_tagrtm_hdr, rxr_long_msgrtm_hdr and long_tagrtm_hdr would not change because of padding.

Reply:
Okay, got it.